include/circt/Support/CPPCDE.h

```diff
@@ -0,0 +1,398 @@
+  using DivU = comb::DivUOp;
+  using DivS = comb::DivSOp;
+  using ModU = comb::ModUOp;
+  using ModS = comb::ModSOp;
+};
+
+// A context containing the shared state of a generator.
+struct CDEContext {
+  CDEContext(Location loc, OpBuilder &b, BackedgeBuilder &bb)
+      : loc(loc), b(b), bb(bb) {}
+  Location loc;
+  OpBuilder &b;
+  BackedgeBuilder &bb;
+  Value clk;
+  Value rst;
+};
+
+// Base class for CPPCDE values. Users can specialize this class for
+// dialect-specific operations.
+template <typename TOps, typename CDEValue>
+class CDEValueImpl {
+
+  template <typename TBinOp>
+  CDEValue execBinOp(CDEValue other) {
+    return CDEValue(this->ctx,
+                    this->ctx->b.template create<TBinOp>(
+                        this->get().getLoc(), this->get(), other.get()));
+  }
+
+public:
+  using OpTypes = TOps;
+  CDEValueImpl() {
+    // Null value.
+  }
+  explicit CDEValueImpl(const CDEValue &other) : ctx(other.ctx) {
+    this->valueOrBackedge = other.valueOrBackedge;
+  }
+  CDEValueImpl(CDEContext *ctx, Value v) : ctx(ctx), valueOrBackedge(v) {}
+  CDEValueImpl(CDEContext *ctx, CDEValue v) : ctx(ctx), valueOrBackedge(v) {}
+  CDEValueImpl(CDEContext *ctx, Backedge *b) : ctx(ctx), valueOrBackedge(b) {}
+
+  virtual ~CDEValueImpl() {}
+
+  bool isNull() { return ctx == nullptr; }
+
+  // Returns a registered version of this value. If no clock is provided,
+  // it is assumed that the context contains a clock value.
+  CDEValue reg(StringRef name, CDEValue clk = CDEValue()) {
+    Value clock;
+    if (clk.isNull()) {
+      assert(ctx->clk && "A clock must be in the CDE context to creae a "
+                         "register.");
+      clock = ctx->clk;
+    } else {
+      clock = clk.get();
+    }
+    return CDEValue(ctx,
+                    ctx->b.create<seq::CompRegOp>(get().getLoc(), get(), clock,
+                                                  ctx->b.getStringAttr(name)));
+  }
+
+  // An explicit function for backedge assignment instead of overloading the
+  // '=' operator.
+  // By this, we avoid aliasing with the implicit copy constructor + don't mix
+  // C++ and the underlying op generation semantics.
+  void assign(CDEValue rhs) {
+    auto *backedge = std::get_if<Backedge *>(&valueOrBackedge);
+    assert(backedge && "Cannot assign to a value.");
+    assert(!(*backedge)->isSet() && "backedge already assigned.");
+    (*backedge)->setValue(rhs.get());
+  }
+
+  // As an alternative to static_cast when implicit conversion won't apply.
+  Value get() const {
+    auto *value = std::get_if<Value>(&valueOrBackedge);
+    if (value)
+      return *value;
+
+    return **std::get_if<Backedge *>(&valueOrBackedge);
+  }
+
+  // Various binary operators.
+  // @todo: These only really make sense for bitvector operations... The dilemma
+  // is to choose between having the user forced to cast values to a CDEValue
+  // specialization (e.g. CDEBitVectorValue) which then contains the operators -
```

Comment:
For the same reasons we do this in PyCDE, I suspect this is the way to go. But it can be added later.